File: translated-python/mathlib.py
"""mathlib.py

Python port of ../original-javascript/mathlib.js, produced by an LLM coding
agent (Claude Code). The port is idiomatic Python rather than a line-by-line
transliteration:

- snake_case names (standardDeviation -> standard_deviation)
- type hints and docstrings
- standard Python exceptions (TypeError, ValueError, ZeroDivisionError)
- a NamedTuple result for linear_regression instead of a plain object
- {sample: false} options objects become a sample=False keyword argument

Numerical behavior is required to match the original exactly. That claim is
enforced, not assumed: test_parity.py replays 86 recorded input/output pairs
produced by executing the original JavaScript, and test_mathlib.py
cross-checks against Python's statistics module. See the README.

Deliberate semantic notes, mirrored from the original:

- Invalid input raises. Functions never return NaN or Infinity. JavaScript
  natively returns Infinity for 1/0 and NaN for sqrt(-1); the original
  library treats both as errors, so this port does too.
- power() uses math.pow rather than the ** operator: in Python,
  (-8) ** 0.5 silently returns a complex number, while the original
  JavaScript treats a non-real result as an error.
- factorial() delegates to math.factorial, which is exact for any n because
  Python integers do not overflow. JavaScript's 64-bit floats overflow above
  170!. The shared, tested domain is n <= 170; behavior above that is
  documented here as an intentional difference (Python keeps working).
- variance/standard_deviation/covariance default to the SAMPLE statistic
  (n - 1 denominator), matching R. Pass sample=False for population.
- quantile uses linear interpolation between order statistics (R type 7,
  NumPy default, statistics.quantiles(method="inclusive")).
"""

from __future__ import annotations

import math
from typing import NamedTuple, Sequence

__all__ = [
    "add",
    "subtract",
    "multiply",
    "divide",
    "power",
    "sqrt",
    "factorial",
    "combinations",
    "mean",
    "median",
    "variance",
    "standard_deviation",
    "quantile",
    "percentile",
    "covariance",
    "correlation",
    "linear_regression",
    "LinearRegressionResult",
]

Number = float | int


# ---------------------------------------------------------------------------
# Input validation helpers
# ---------------------------------------------------------------------------


def _check_number(value: Number, name: str) -> None:
    if isinstance(value, bool) or not isinstance(value, (int, float)):
        raise TypeError(f"{name} must be a finite number")
    if not math.isfinite(value):
        raise TypeError(f"{name} must be a finite number")


def _check_number_sequence(values: Sequence[Number], name: str, min_length: int) -> None:
    if isinstance(values, (str, bytes)) or not isinstance(values, Sequence):
        raise TypeError(f"{name} must be a sequence of numbers")
    if len(values) < min_length:
        raise ValueError(f"{name} must contain at least {min_length} element(s)")
    for v in values:
        _check_number(v, f"every element of {name}")


def _check_same_length(xs: Sequence[Number], ys: Sequence[Number]) -> None:
    if len(xs) != len(ys):
        raise ValueError("xs and ys must have the same length")


# ---------------------------------------------------------------------------
# Basic arithmetic
# ---------------------------------------------------------------------------


def add(a: Number, b: Number) -> float:
    """Return a + b."""
    _check_number(a, "a")
    _check_number(b, "b")
    return a + b


def subtract(a: Number, b: Number) -> float:
    """Return a - b."""
    _check_number(a, "a")
    _check_number(b, "b")
    return a - b


def multiply(a: Number, b: Number) -> float:
    """Return a * b."""
    _check_number(a, "a")
    _check_number(b, "b")
    return a * b


def divide(a: Number, b: Number) -> float:
    """Return a / b. Raises ZeroDivisionError when b is 0."""
    _check_number(a, "a")
    _check_number(b, "b")
    if b == 0:
        raise ZeroDivisionError("division by zero")
    return a / b


def power(base: Number, exponent: Number) -> float:
    """Return base raised to exponent.

    Raises ValueError when the result is not a finite real number, e.g.
    power(0, -1) or power(-8, 0.5). math.pow is used instead of ** so that
    a negative base with a fractional exponent is an error, as in the
    original, rather than a complex number.
    """
    _check_number(base, "base")
    _check_number(exponent, "exponent")
    try:
        result = math.pow(base, exponent)
    except (ValueError, OverflowError) as exc:
        raise ValueError("power: result is not a finite real number") from exc
    if not math.isfinite(result):
        raise ValueError("power: result is not a finite real number")
    return result


def sqrt(x: Number) -> float:
    """Return the square root of x. Raises ValueError for negative x."""
    _check_number(x, "x")
    if x < 0:
        raise ValueError("sqrt: x must be non-negative")
    return math.sqrt(x)


def factorial(n: Number) -> int:
    """Return n! for a non-negative integer n.

    Raises ValueError for negative or non-integer n. Unlike the JavaScript
    original, the result is an exact integer for any n (Python integers do
    not overflow); the original overflows above n = 170.
    """
    _check_number(n, "n")
    if isinstance(n, float):
        if not n.is_integer():
            raise ValueError("factorial: n must be a non-negative integer")
        n = int(n)
    if n < 0:
        raise ValueError("factorial: n must be a non-negative integer")
    return math.factorial(n)


def combinations(n: Number, k: Number) -> int:
    """Return the binomial coefficient C(n, k).

    Raises ValueError unless n and k are integers with 0 <= k <= n.
    """
    _check_number(n, "n")
    _check_number(k, "k")
    if isinstance(n, float):
        if not n.is_integer():
            raise ValueError("combinations: need integers with 0 <= k <= n")
        n = int(n)
    if isinstance(k, float):
        if not k.is_integer():
            raise ValueError("combinations: need integers with 0 <= k <= n")
        k = int(k)
    if n < 0 or k < 0 or k > n:
        raise ValueError("combinations: need integers with 0 <= k <= n")
    return math.comb(n, k)


# ---------------------------------------------------------------------------
# Descriptive statistics
# ---------------------------------------------------------------------------


def mean(values: Sequence[Number]) -> float:
    """Return the arithmetic mean of a non-empty sequence."""
    _check_number_sequence(values, "values", 1)
    return sum(values) / len(values)


def median(values: Sequence[Number]) -> float:
    """Return the median of a non-empty sequence."""
    _check_number_sequence(values, "values", 1)
    ordered = sorted(values)
    mid = len(ordered) // 2
    if len(ordered) % 2 == 1:
        return ordered[mid]
    return (ordered[mid - 1] + ordered[mid]) / 2


def variance(values: Sequence[Number], *, sample: bool = True) -> float:
    """Return the variance of a sequence.

    Sample variance (n - 1 denominator) by default, matching R; pass
    sample=False for population variance. Sample variance requires at
    least 2 values.
    """
    _check_number_sequence(values, "values", 2 if sample else 1)
    m = mean(values)
    sum_sq = 0.0
    for v in values:
        sum_sq += (v - m) * (v - m)
    return sum_sq / (len(values) - (1 if sample else 0))


def standard_deviation(values: Sequence[Number], *, sample: bool = True) -> float:
    """Return the standard deviation: the square root of variance()."""
    return math.sqrt(variance(values, sample=sample))


def quantile(values: Sequence[Number], q: Number) -> float:
    """Return the q-th quantile (0 <= q <= 1) of a non-empty sequence.

    Uses linear interpolation between order statistics (R type 7, the
    NumPy default). Raises ValueError when q is outside [0, 1].
    """
    _check_number_sequence(values, "values", 1)
    _check_number(q, "q")
    if q < 0 or q > 1:
        raise ValueError("quantile: q must be between 0 and 1")
    ordered = sorted(values)
    position = (len(ordered) - 1) * q
    lower = math.floor(position)
    upper = math.ceil(position)
    if lower == upper:
        return ordered[lower]
    return ordered[lower] + (ordered[upper] - ordered[lower]) * (position - lower)


def percentile(values: Sequence[Number], p: Number) -> float:
    """Return the p-th percentile (0 <= p <= 100): quantile(values, p / 100)."""
    _check_number(p, "p")
    if p < 0 or p > 100:
        raise ValueError("percentile: p must be between 0 and 100")
    return quantile(values, p / 100)


# ---------------------------------------------------------------------------
# Bivariate statistics
# ---------------------------------------------------------------------------


def covariance(
    xs: Sequence[Number], ys: Sequence[Number], *, sample: bool = True
) -> float:
    """Return the covariance of two equal-length sequences.

    Sample covariance (n - 1 denominator) by default; pass sample=False
    for the population statistic.
    """
    min_length = 2 if sample else 1
    _check_number_sequence(xs, "xs", min_length)
    _check_number_sequence(ys, "ys", min_length)
    _check_same_length(xs, ys)
    mx = mean(xs)
    my = mean(ys)
    total = 0.0
    for x, y in zip(xs, ys):
        total += (x - mx) * (y - my)
    return total / (len(xs) - (1 if sample else 0))


def correlation(xs: Sequence[Number], ys: Sequence[Number]) -> float:
    """Return the Pearson correlation coefficient of two sequences.

    Raises ValueError when either sequence is constant (undefined).
    """
    cov = covariance(xs, ys, sample=True)
    sx = standard_deviation(xs, sample=True)
    sy = standard_deviation(ys, sample=True)
    if sx == 0 or sy == 0:
        raise ValueError("correlation: undefined for constant input")
    return cov / (sx * sy)


class LinearRegressionResult(NamedTuple):
    """Ordinary least squares fit: y = intercept + slope * x."""

    slope: float
    intercept: float
    r2: float


def linear_regression(
    xs: Sequence[Number], ys: Sequence[Number]
) -> LinearRegressionResult:
    """Fit ordinary least squares y = intercept + slope * x.

    Raises ValueError when xs is constant. When ys is constant the fit is
    exact and r2 is defined as 1.
    """
    var_x = variance(xs, sample=True)
    if var_x == 0:
        raise ValueError("linear_regression: xs must not be constant")
    slope = covariance(xs, ys, sample=True) / var_x
    intercept = mean(ys) - slope * mean(xs)
    my = mean(ys)
    ss_res = 0.0
    ss_tot = 0.0
    for x, y in zip(xs, ys):
        fitted = intercept + slope * x
        ss_res += (y - fitted) * (y - fitted)
        ss_tot += (y - my) * (y - my)
    r2 = 1.0 if ss_tot == 0 else 1 - ss_res / ss_tot
    return LinearRegressionResult(slope=slope, intercept=intercept, r2=r2)

File: original-javascript/mathlib.js
/**
 * mathlib.js
 *
 * A small arithmetic and statistics library in plain JavaScript (Node.js).
 *
 * This is the ORIGINAL source in the translation demo. The Python port in
 * ../translated-python/mathlib.py was produced from this file by an LLM
 * coding agent, and the test suite verifies that the two implementations
 * behave identically.
 *
 * Design decisions (the Python port must preserve these):
 *
 * - Invalid input raises an error. Functions never return NaN or Infinity.
 *   (Plain JavaScript would happily return Infinity for 1/0 and NaN for
 *   Math.sqrt(-1); this library treats both as errors.)
 * - variance, standardDeviation, and covariance compute the SAMPLE statistic
 *   by default (n - 1 denominator), matching R. Pass {sample: false} for the
 *   population statistic (n denominator).
 * - quantile uses linear interpolation between order statistics, the same
 *   method as R's default (type 7) and NumPy's default.
 */

'use strict';

// ---------------------------------------------------------------------------
// Input validation helpers
// ---------------------------------------------------------------------------

function assertFiniteNumber(value, name) {
  if (typeof value !== 'number' || !Number.isFinite(value)) {
    throw new TypeError(name + ' must be a finite number');
  }
}

function assertNumberArray(values, name, minLength) {
  if (!Array.isArray(values)) {
    throw new TypeError(name + ' must be an array of numbers');
  }
  if (values.length < minLength) {
    throw new RangeError(name + ' must contain at least ' + minLength + ' element(s)');
  }
  for (const v of values) {
    assertFiniteNumber(v, 'every element of ' + name);
  }
}

function assertSameLength(xs, ys) {
  if (xs.length !== ys.length) {
    throw new RangeError('xs and ys must have the same length');
  }
}

// ---------------------------------------------------------------------------
// Basic arithmetic
// ---------------------------------------------------------------------------

/** Returns a + b. */
function add(a, b) {
  assertFiniteNumber(a, 'a');
  assertFiniteNumber(b, 'b');
  return a + b;
}

/** Returns a - b. */
function subtract(a, b) {
  assertFiniteNumber(a, 'a');
  assertFiniteNumber(b, 'b');
  return a - b;
}

/** Returns a * b. */
function multiply(a, b) {
  assertFiniteNumber(a, 'a');
  assertFiniteNumber(b, 'b');
  return a * b;
}

/**
 * Returns a / b.
 * Throws RangeError when b is 0. (Plain JavaScript would return Infinity.)
 */
function divide(a, b) {
  assertFiniteNumber(a, 'a');
  assertFiniteNumber(b, 'b');
  if (b === 0) {
    throw new RangeError('division by zero');
  }
  return a / b;
}

/**
 * Returns base raised to exponent.
 * Throws RangeError when the result is not a finite real number, e.g.
 * power(0, -1) (division by zero) or power(-8, 0.5) (complex result).
 */
function power(base, exponent) {
  assertFiniteNumber(base, 'base');
  assertFiniteNumber(exponent, 'exponent');
  const result = Math.pow(base, exponent);
  if (!Number.isFinite(result)) {
    throw new RangeError('power: result is not a finite real number');
  }
  return result;
}

/**
 * Returns the square root of x.
 * Throws RangeError for negative x. (Math.sqrt would return NaN.)
 */
function sqrt(x) {
  assertFiniteNumber(x, 'x');
  if (x < 0) {
    throw new RangeError('sqrt: x must be non-negative');
  }
  return Math.sqrt(x);
}

/**
 * Returns n! for a non-negative integer n.
 * Throws RangeError for negative or non-integer n, and for n large enough
 * that the result overflows a 64-bit float (n > 170).
 */
function factorial(n) {
  assertFiniteNumber(n, 'n');
  if (!Number.isInteger(n) || n < 0) {
    throw new RangeError('factorial: n must be a non-negative integer');
  }
  let result = 1;
  for (let i = 2; i <= n; i++) {
    result *= i;
  }
  if (!Number.isFinite(result)) {
    throw new RangeError('factorial: result overflows a 64-bit float');
  }
  return result;
}

/**
 * Returns the binomial coefficient C(n, k): the number of ways to choose
 * k items from n. Throws RangeError unless n and k are integers with
 * 0 <= k <= n.
 */
function combinations(n, k) {
  assertFiniteNumber(n, 'n');
  assertFiniteNumber(k, 'k');
  if (!Number.isInteger(n) || !Number.isInteger(k) || n < 0 || k < 0 || k > n) {
    throw new RangeError('combinations: need integers with 0 <= k <= n');
  }
  if (k > n - k) {
    k = n - k;
  }
  let result = 1;
  for (let i = 1; i <= k; i++) {
    result = (result * (n - k + i)) / i;
  }
  return Math.round(result);
}

// ---------------------------------------------------------------------------
// Descriptive statistics
// ---------------------------------------------------------------------------

/** Returns the arithmetic mean of a non-empty array. */
function mean(values) {
  assertNumberArray(values, 'values', 1);
  let total = 0;
  for (const v of values) {
    total += v;
  }
  return total / values.length;
}

/**
 * Returns the median of a non-empty array.
 * Note the comparator passed to sort(): JavaScript's default sort is
 * LEXICOGRAPHIC, so [10, 9].sort() stays [10, 9]. A correct numeric sort
 * must be requested explicitly.
 */
function median(values) {
  assertNumberArray(values, 'values', 1);
  const sorted = [...values].sort((a, b) => a - b);
  const mid = Math.floor(sorted.length / 2);
  if (sorted.length % 2 === 1) {
    return sorted[mid];
  }
  return (sorted[mid - 1] + sorted[mid]) / 2;
}

/**
 * Returns the variance of an array.
 * Sample variance (n - 1 denominator) by default; pass {sample: false}
 * for population variance. Sample variance requires at least 2 values.
 */
function variance(values, { sample = true } = {}) {
  assertNumberArray(values, 'values', sample ? 2 : 1);
  const m = mean(values);
  let sumSq = 0;
  for (const v of values) {
    sumSq += (v - m) * (v - m);
  }
  return sumSq / (values.length - (sample ? 1 : 0));
}

/** Returns the standard deviation: the square root of variance(). */
function standardDeviation(values, options) {
  return Math.sqrt(variance(values, options));
}

/**
 * Returns the q-th quantile (0 <= q <= 1) of a non-empty array, using
 * linear interpolation between order statistics (R type 7, NumPy default).
 */
function quantile(values, q) {
  assertNumberArray(values, 'values', 1);
  assertFiniteNumber(q, 'q');
  if (q < 0 || q > 1) {
    throw new RangeError('quantile: q must be between 0 and 1');
  }
  const sorted = [...values].sort((a, b) => a - b);
  const position = (sorted.length - 1) * q;
  const lower = Math.floor(position);
  const upper = Math.ceil(position);
  if (lower === upper) {
    return sorted[lower];
  }
  return sorted[lower] + (sorted[upper] - sorted[lower]) * (position - lower);
}

/** Returns the p-th percentile (0 <= p <= 100). Equal to quantile(values, p / 100). */
function percentile(values, p) {
  assertFiniteNumber(p, 'p');
  if (p < 0 || p > 100) {
    throw new RangeError('percentile: p must be between 0 and 100');
  }
  return quantile(values, p / 100);
}

// ---------------------------------------------------------------------------
// Bivariate statistics
// ---------------------------------------------------------------------------

/**
 * Returns the covariance of two equal-length arrays.
 * Sample covariance (n - 1 denominator) by default; pass {sample: false}
 * for the population statistic.
 */
function covariance(xs, ys, { sample = true } = {}) {
  assertNumberArray(xs, 'xs', sample ? 2 : 1);
  assertNumberArray(ys, 'ys', sample ? 2 : 1);
  assertSameLength(xs, ys);
  const mx = mean(xs);
  const my = mean(ys);
  let total = 0;
  for (let i = 0; i < xs.length; i++) {
    total += (xs[i] - mx) * (ys[i] - my);
  }
  return total / (xs.length - (sample ? 1 : 0));
}

/**
 * Returns the Pearson correlation coefficient of two equal-length arrays.
 * Throws RangeError when either array is constant (correlation undefined).
 */
function correlation(xs, ys) {
  const cov = covariance(xs, ys, { sample: true });
  const sx = standardDeviation(xs, { sample: true });
  const sy = standardDeviation(ys, { sample: true });
  if (sx === 0 || sy === 0) {
    throw new RangeError('correlation: undefined for constant input');
  }
  return cov / (sx * sy);
}

/**
 * Fits ordinary least squares y = intercept + slope * x.
 * Returns {slope, intercept, r2}. Throws RangeError when xs is constant.
 * When ys is constant the fit is exact and r2 is defined as 1.
 */
function linearRegression(xs, ys) {
  const varX = variance(xs, { sample: true });
  if (varX === 0) {
    throw new RangeError('linearRegression: xs must not be constant');
  }
  const slope = covariance(xs, ys, { sample: true }) / varX;
  const intercept = mean(ys) - slope * mean(xs);
  const my = mean(ys);
  let ssRes = 0;
  let ssTot = 0;
  for (let i = 0; i < xs.length; i++) {
    const fitted = intercept + slope * xs[i];
    ssRes += (ys[i] - fitted) * (ys[i] - fitted);
    ssTot += (ys[i] - my) * (ys[i] - my);
  }
  const r2 = ssTot === 0 ? 1 : 1 - ssRes / ssTot;
  return { slope, intercept, r2 };
}

module.exports = {
  add,
  subtract,
  multiply,
  divide,
  power,
  sqrt,
  factorial,
  combinations,
  mean,
  median,
  variance,
  standardDeviation,
  quantile,
  percentile,
  covariance,
  correlation,
  linearRegression,
};
